Ellipsis fitting: Check the condition number before computiing eigen vectors

hmaarrfk opened the pull request:
Under certain conditions, the result of a matrix we have to invert becomes ill conditioneed. The 32 bit solver and 64 bit solver converge on different answers.
We first check if that matrix is ill conditioned before we move forward with the computation.

See #3091

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]


## References
[If this is a bug-fix or enhancement, it closes issue # ]
[If this is a new feature, it implements the following paper: ]

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

pep8speaks commented:
Hello @hmaarrfk! Thanks for submitting the PR.

 - In the file [`skimage/measure/fit.py`](https://github.com/scikit-image/scikit-image/blob/44aeea4e5713719801961ad06cf43eeafb1740dc/skimage/measure/fit.py), following are the PEP8 issues :

> [Line 439:33](https://github.com/scikit-image/scikit-image/blob/44aeea4e5713719801961ad06cf43eeafb1740dc/skimage/measure/fit.py#L439): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator

stefanv commented on skimage/measure/fit.py:
```diff
@@ -434,6 +434,11 @@ def estimate(self, data):
 
+        # For some reason the above doesn't catch the error on linux 32 bit
+        # https://stackoverflow.com/questions/13249108/efficient-pythonic-check-for-singular-matrix
+        if np.linalg.cond(M) > 1 / np.finfo(np.double).eps:
```
This check can be done before the inversion is attempted?  It simplifies the logic a bit.

hmaarrfk replied:
I'm not sure. in fact, when I computed `cond(M)` in the debugger with `import pdb; pdb.set_trace()`, the condition number was already `Inf`. I don't really want to cause other overflow errors or "old numpy specific errors"

From the docs
> eps | (float) The smallest representable positive number such that 1.0 + eps != 1.0.  Type of eps is an appropriate floating point type.

They don't seem to have a eps^-1. Thought I could do
`eps**-1` if you like.

stefanv commented:
It seems like the test suite already fails, but it may be worth inserting an explicit test for this, potentially one that also fails on 64-bit.

matthew-brett commented:
Should this be a bug report to numpy too?

stefanv commented:
@matthew-brett I've raised this with the NumPy folks; they think it's not NumPy's responsibility.

Overall, it is a bad idea to use inverse if you can use solve.  Could we rewrite the solution here to do solves instead?

hmaarrfk commented:
Just for you two, I actually stepped through to figure out what was happening.

`S1` and `C1` are well conditioned. It is the result of the matrix multiplications and addition that is illconditionned. Typically this is caught a little lower

```python
        # M*|a b c >=l|a b c >. Find eigenvalues and eigenvectors
        # from this equation [eqn. 28]
        eig_vals, eig_vecs = np.linalg.eig(M)

        # eigenvector must meet constraint 4ac - b^2 to be valid.
        cond = 4 * np.multiply(eig_vecs[0, :], eig_vecs[2, :]) \
               - np.power(eig_vecs[1, :], 2)
        a1 = eig_vecs[:, (cond > 0)]
        # seeks for empty matrix
        if 0 in a1.shape or len(a1.ravel()) != 3:
            return False
```
On 64 bit:
```
(Pdb) cond
array([-0.28717554, -0.71271829, -0.00108459])
```
On 32 bit
```
(Pdb) cond
array([-1.38708836,  0.84396406, -0.03172746])
```

It just happened that on 64 bit the error didn't show up. But I don't think that the `eig` algorithm will complain if you have a ill-conditionned matrix.

Therefore, the tests both in *64 bit* and *32 bit* are caught by my added logic.

hmaarrfk commented:
Technically, since the matrix is so small, I think we can simply check that if one `eig_vals` is small and also return `False`.

Anyway, I think checking the condition number is more "mathematical"

hmaarrfk commented:
@stefanv. I probably agree that doing something like finding tge best eigen value is probably verbose. 

From ‘numpy.linalg.solve’

> a must be square and of full-rank 

We would need to check the condition number of the matrix anyway otherwise risk having an other linalg error

hmaarrfk commented:
Finally @stefanv. I quickly skimmed the paper and the code seems to be following the formalism the paper is using.

I suggest that we leave the reformulation to somebody that wants to derive it. (It May be a paper for itself)

emmanuelle commented:
So, should we merge this PR?

hmaarrfk commented:
I think so! I can rebase if you want.

jni commented:
@hmaarrfk I can't quite tell from the conversation whether you have a test case that will fail on 64 bit? It would be nice for this to be tested.

hmaarrfk commented:
Sorry, I guess I coudln't think of one. I'll try to bring out my pen and paper this weekend. Remind me.

hmaarrfk committed:
TST: Add a 64 bit failer for ellipsis fitting

hmaarrfk commented:
@jni 

```
=============================================== FAILURES ================================================
__________________________________ test_ellipse_model_estimate_failers __________________________________

    def test_ellipse_model_estimate_failers():
        # estimate parameters of real data
        model = EllipseModel()
        assert not model.estimate(np.ones((5, 2)))
        # Before PR https://github.com/scikit-image/scikit-image/pull/3103
        # This next line would return True on 32bit
        assert not model.estimate(np.array([[50, 80], [51, 81], [52, 80]]))
        # This is the same test that causes the model to
        # return true on 64 bit prooving that this was also an issue on
        # 64 bit arch
>       assert not model.estimate(np.array([[50, 80], [51, 81], [52, 80.00000000001]]))
E       assert not True
E        +  where True = <bound method EllipseModel.estimate of <skimage.measure.fit.EllipseModel object at 0x7f422d07e5c0>>(array([[ 50.,  80.],\n       [ 51.,  81.],\n       [ 52.,  80.]]))
E        +    where <bound method EllipseModel.estimate of <skimage.measure.fit.EllipseModel object at 0x7f422d07e5c0>> = <skimage.measure.fit.EllipseModel object at 0x7f422d07e5c0>.estimate
E        +    and   array([[ 50.,  80.],\n       [ 51.,  81.],\n       [ 52.,  80.]]) = <built-in function array>([[50, 80], [51, 81], [52, 80.00000000001]])
E        +      where <built-in function array> = np.array

skimage/measure/tests/test_fit.py:233: AssertionError
================================== 1 failed, 22 passed in 0.72 seconds ==================================
```

You can go ahead and force push to remove my "Reverting" of the fix once travis shows that the tests fail. I'm not really a fan of using `eps`. Apparently it is pretty slow the first time you call it.

Also, you can precompute the inv of `C1`, and paste it in instead. It is nice round numbers so it stays readable. I figured that adding that in would delay this PR getting merged.

hmaarrfk commented:
Failed on Appveyor: https://ci.appveyor.com/project/scikit-image/scikit-image/build/1.0.1014/job/qqf67dhmccr8f0so#L8458
Failed on Travis python 3.6: https://travis-ci.org/scikit-image/scikit-image/jobs/430813979#L2803
But passed on python 3.7: https://travis-ci.org/scikit-image/scikit-image/jobs/430813978

hmaarrfk committed:
Fixes #3091: check for condition number before during ellipsis fitting

Explicitely check the condition number of the matrix before computing
the eigen vectors

hmaarrfk commented:
@jni Here is what recreates the issue:

```python
from skimage.measure import EllipseModel
import numpy as np
model = EllipseModel()
print(model.estimate(np.array([[50, 80], [51, 81], [52, 80.00000000001]])))
print(model.estimate(np.array([[50, 80], [51, 81], [52, 80.0001000001]])))
print(model.estimate(np.array([[50, 80], [51, 81], [52, 80.000000001]])))
print(model.estimate(np.array([[50, 80], [51, 81], [52, 80.00000001]])))
print(model.estimate(np.array([[50, 80], [51, 81], [52, 80.0000001]])))
print(model.estimate(np.array([[50, 80], [51, 81], [52, 80.000001]])))
print(model.estimate(np.array([[50, 80], [51, 81], [52, 80.00001]])))
```

```
False
False
True
True
True
True
False
```
I thought you could simply check the condition number, but it fails the doctest, which should clearly pass....

hmaarrfk commented:
part of me thinks we should chek the number of input points. The problem of fitting an ellipse is ilposed with less than 5 points

jni commented on skimage/measure/tests/test_fit.py:
```diff
@@ -229,7 +229,22 @@ def test_ellipse_model_estimate_failers():
+    # 64 bit arch
+    assert not model.estimate(np.array([[50, 80], [51, 81], [52, 80.00000000001]]))
+    # Unfortunately, it seems to depend on the version of python that is used
+    # Therefore, test for many cases
```
@hmaarrfk based on your comment, shouldn't this test be:

```python
assert not all(model.estimate(np.array([[50, 80], [51, 81], [52, 80 + 10**(-n)]]))
               for n in range(4, 12))
```
?

hmaarrfk replied:
@jni, refactored as per your suggestion. problem remains :/

jni replied:
Which problem? Do they all pass for some Python versions?

hmaarrfk replied:
Well my **fix** causes the doctest to fail. But the doctest seems to make sense. https://travis-ci.org/scikit-image/scikit-image/jobs/432304643#L2823

I'm glad to make the test xfail for code self-documentation.

jni replied:
OOOOoooooOOOOhhhh. Very interesting. Hmm, I think ellipse-fitting to a line should work, actually. =\

Borda commented on skimage/measure/tests/test_fit.py:
```diff
@@ -225,11 +225,20 @@ def test_ellipse_model_estimate_from_data():
-    assert not model.estimate(np.array([[50, 80], [51, 81], [52, 80]]))
+    assert not model.estimate(np.array([[50, 80],
+                                        [51, 81],
+                                        [52, 80 + epsilon]]))
```
according to the comment, what is the epsilon number (it is not clear from the review scope)

hmaarrfk replied:
sorry, quite often, epsilon is a small number in math. I could add a comment, but this wasn't the right approach anyway, and gave the same results as before.

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
